diff --git a/arch/Kconfig b/arch/Kconfig
index 9003a6e6f3d1..f4f559e0d413 100644
--- a/arch/Kconfig
+++ b/arch/Kconfig
@@ -739,6 +739,10 @@ config CFI_PERMISSIVE
 	  warning instead of a kernel panic. This option is useful for finding
 	  CFI violations during development.
 
+config CFI_FORCE_SKIP_CHECK
+	bool "Skip execution of module __cfi_check functions"
+	depends on CFI_CLANG
+
 config HAVE_ARCH_WITHIN_STACK_FRAMES
 	bool
 	help
diff --git a/kernel/cfi.c b/kernel/cfi.c
index 9594cfd1cf2c..7fbfef34fca1 100644
--- a/kernel/cfi.c
+++ b/kernel/cfi.c
@@ -323,9 +323,11 @@ void cfi_slowpath_handler(uint64_t id, void *ptr, void *diag)
 	if (!IS_ENABLED(CONFIG_CFI_PERMISSIVE))
 		diag = NULL;
 
-	if (likely(fn))
+	if (likely(fn)) {
+#ifndef CONFIG_CFI_FORCE_SKIP_CHECK
 		fn(id, ptr, diag);
-	else /* Don't allow unchecked modules */
+#endif
+	} else /* Don't allow unchecked modules */
 		handle_cfi_failure(ptr);
 }
